File: ops/ascendc/torch_ops_extension/custom_ops/converter/npu_sparse_flash_attention_antiquant.py
# Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import (
    Any, Callable, ContextManager, Iterator, List, Literal, NamedTuple, Optional, Sequence, Tuple, TypeVar,
    Union, overload,
)

import torch
import torchair
from torch import Generator, contiguous_format, inf, strided, SymInt
from torch.types import Device, Number, _bool, _complex, _device, _dtype, _float, _int, _layout, _qscheme, _size
from torchair._ge_concrete_graph import ge_apis as ge
from torchair._ge_concrete_graph.fx2ge_converter import declare_supported, register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec, DataType
from torchair._ge_concrete_graph.supported_declaration import _TypedTensor, F32, F16, F64, I32, I16, I64, I8, U8, \
    BOOL, Support
from torchair.ge import attr


# 为自定义算子注册converter，用于torch.compile 场景成图
# 注意： meta_outputs形参名为固定写法，若写错会影响ge节点的输出dtype与shape推导
@register_fx_node_ge_converter(torch.ops.custom.npu_sparse_flash_attention_antiquant.default)
def convert_npu_sparse_flash_attention_antiquant(
    query: Tensor,
    key: Tensor,
    value: Tensor,
    sparse_indices: Tensor,
    scale_value: float,
    sparse_block_size: int,
    key_quant_mode: int,
    value_quant_mode: int,
    *,
    key_dequant_scale: Optional[Tensor] = None,
    value_dequant_scale: Optional[Tensor] = None,
    block_table: Optional[Tensor] = None,
    actual_seq_lengths_query: Optional[Tensor] = None,
    actual_seq_lengths_kv: Optional[Tensor] = None,
    layout_query: str = "BSND",
    layout_kv: str = "BSND",
    sparse_mode: int = 3,
    attention_mode: int = 0,
    quant_scale_repo_mode: int = 0,
    tile_size: int = 0,
    rope_head_dim: int = 0,
    meta_outputs: TensorSpec = None,
):
    return torchair.ge.custom_op(
        "SparseFlashAttentionAntiquant",
        inputs={"query": query, 
                "key": key,
                "value": value,
                "sparse_indices": sparse_indices,
                "key_dequant_scale": key_dequant_scale,
                "value_dequant_scale": value_dequant_scale,
                "block_table": block_table,
                "actual_seq_lengths_query": actual_seq_lengths_query,
                "actual_seq_lengths_kv": actual_seq_lengths_kv,
               },
        attrs={"scale_value": attr.Float(scale_value),
               "sparse_block_size": attr.Int(sparse_block_size),
               "key_quant_mode": attr.Int(key_quant_mode),
               "value_quant_mode": attr.Int(value_quant_mode),
               "layout_query": attr.Str(layout_query),
               "layout_kv": attr.Str(layout_kv),
               "sparse_mode": attr.Int(sparse_mode),
               "attention_mode": attr.Int(attention_mode),
               "quant_scale_repo_mode": attr.Int(quant_scale_repo_mode),
               "tile_size": attr.Int(tile_size),
               "rope_head_dim": attr.Int(rope_head_dim),
               },
        outputs=['attention_out']
    )

File: ops/ascendc/torch_ops_extension/custom_ops/csrc/npu_sparse_flash_attention_antiquant.cpp
/* *
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <iostream>
#include <torch/library.h>
#include "ops_common.h"

namespace custom {
using namespace at_npu::native;

// npu tensor max size
const int SIZE = 8;
const int DIM_0 = 0;
const int DIM_1 = 1;
const int DIM_2 = 2;
const int DIM_3 = 3;

// 工具函数，推导输出shape
at::Tensor construct_sparse_antiquant_infer_output_tensor(
    const at::Tensor& query, const at::Tensor& value, std::string layout_query_str,
    std::string layout_kv_str, const uint64_t &rope_head_dim)
{
    for (auto i = 0; i < query.sizes().size(); i++) {
        TORCH_CHECK(query.size(i) > 0, "All values within query's shape should be greater "
            "than 0, but shape[", i, "] is ", query.size(i));
    }
    at::SmallVector<int64_t, SIZE> output_size;
    if (layout_query_str == "BSND") {
        output_size = {query.size(DIM_0), query.size(DIM_1), query.size(DIM_2), query.size(DIM_3) - rope_head_dim};
    } else {
        output_size = {query.size(DIM_0), query.size(DIM_1), query.size(DIM_2) - rope_head_dim};
    }
    at::Tensor output = at::empty(output_size, query.options().dtype(query.dtype()));

    return output;
}

// step2, 为NPU设备实现前向接口
at::Tensor npu_sparse_flash_attention_antiquant_npu(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const at::Tensor &sparse_indices, double scale_value, int64_t sparse_block_size,
    int64_t key_quant_mode, int64_t value_quant_mode,
    const c10::optional<at::Tensor> &key_dequant_scale,
    const c10::optional<at::Tensor> &value_dequant_scale,
    const c10::optional<at::Tensor> &block_table,
    const c10::optional<at::Tensor> &actual_seq_lengths_query,
    const c10::optional<at::Tensor> &actual_seq_lengths_kv,
    c10::string_view layout_query, c10::string_view layout_kv,
    int64_t sparse_mode, int64_t attention_mode, int64_t quant_scale_repo_mode,
    int64_t tile_size, int64_t rope_head_dim)
{
    std::string layout_query_str = std::string(layout_query);
    std::string layout_kv_str = std::string(layout_kv);

    // construct the output tensor
    at::Tensor output = construct_sparse_antiquant_infer_output_tensor(
        query, value, layout_query_str, layout_kv_str, rope_head_dim);
    // convert str
    char *layout_query_ptr = const_cast<char *>(layout_query_str.c_str());
    char *layout_kv_ptr = const_cast<char *>(layout_kv_str.c_str());

    EXEC_NPU_CMD_V1(aclnnSparseFlashAttentionAntiquant, query,
        key, value, sparse_indices, key_dequant_scale, value_dequant_scale, block_table, actual_seq_lengths_query,
        actual_seq_lengths_kv, scale_value, sparse_block_size, key_quant_mode, value_quant_mode,
        layout_query_ptr, layout_kv_ptr, sparse_mode, attention_mode, quant_scale_repo_mode, tile_size, rope_head_dim,
        output);
    return output;
}

// step3, 为META设备实现前向接口
at::Tensor npu_sparse_flash_attention_antiquant_meta(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const at::Tensor &sparse_indices, double scale_value, int64_t sparse_block_size,
    int64_t key_quant_mode, int64_t value_quant_mode,
    const c10::optional<at::Tensor> &key_dequant_scale,
    const c10::optional<at::Tensor> &value_dequant_scale,
    const c10::optional<at::Tensor> &block_table,
    const c10::optional<at::Tensor> &actual_seq_lengths_query,
    const c10::optional<at::Tensor> &actual_seq_lengths_kv,
    c10::string_view layout_query, c10::string_view layout_kv,
    int64_t sparse_mode, int64_t attention_mode, int64_t quant_scale_repo_mode,
    int64_t tile_size, int64_t rope_head_dim)
{
    std::string layout_query_str = std::string(layout_query);
    std::string layout_kv_str = std::string(layout_kv);
    at::Tensor output = construct_sparse_antiquant_infer_output_tensor(
        query, value, layout_query_str, layout_kv_str, rope_head_dim);

    return output;
}
}

// step4, 为NPU设备注册前向实现
TORCH_LIBRARY_IMPL(custom, PrivateUse1, m) {
    m.impl("npu_sparse_flash_attention_antiquant", &custom::npu_sparse_flash_attention_antiquant_npu);
}

// step5, 为META设备注册前向实现
TORCH_LIBRARY_IMPL(custom, Meta, m) {
    m.impl("npu_sparse_flash_attention_antiquant", &custom::npu_sparse_flash_attention_antiquant_meta);
}
